presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java

```diff
@@ -2569,6 +2571,89 @@ public void testMaterializedViewForJoinWithMultiplePartitions()
+                    "SELECT orderkey, linenumber FROM lineitem", table1));
+            queryRunner.execute(format("CREATE TABLE %s WITH (partitioned_by = ARRAY['orderstatus']) AS " +
+                    "SELECT orderkey, custkey, orderstatus FROM orders", table2));
+            assertUpdate(format(
+                    "CREATE MATERIALIZED VIEW %s " +
+                            "WITH (partitioned_by = ARRAY['orderstatus', 'linenumber'])" +
+                            "AS %s",
+                    viewName,
+                    joinSelect));
+
+            final MaterializedResult baseResult = computeActual(joinSelect);
+            final MaterializedResult viewResult = computeActual(format("SELECT * FROM %s", viewName));
+
+            Assert.assertEquals(baseResult, viewResult);
+        }
+        finally {
+            assertUpdate(format("DROP MATERIALIZED VIEW IF EXISTS %s", viewName));
+            queryRunner.execute(format("DROP TABLE IF EXISTS %s", table1));
+            queryRunner.execute(format("DROP TABLE IF EXISTS %s", table2));
+        }
+    }
+
+    @Test
+    public void testMaterializedViewResultForThreeTableJoin()
+    {
+        final String viewName = "mv";
+        final String table1 = "lineitem_partitioned";
+        final String table2 = "orders_partitioned";
+        final String table3 = "customer_partitioned";
+
+        @Language("SQL") final String joinSelect = format("SELECT %s.custkey, %s.orderkey, %s.nationkey, orderstatus, linenumber " +
+                        "FROM %s JOIN %s ON " +
+                        "%s.orderkey = %s.orderkey " +
+                        "JOIN %s ON " +
+                        "%s.custkey = %s.custkey " +
+                        "ORDER BY custkey, orderkey, nationkey, orderstatus, linenumber ",
+                table2, table2, table3,
+                table1, table2,
+                table1, table2,
+                table3,
+                table2, table3);
+        @Language("SQL") final String createMV = format(
+                "CREATE MATERIALIZED VIEW %s " +
+                        "WITH (partitioned_by = ARRAY['nationkey', 'orderstatus', 'linenumber'])" +
+                        "AS %s",
+                viewName,
+                joinSelect);
+        QueryRunner queryRunner = getQueryRunner();
+        try {
+            queryRunner.execute(format("CREATE TABLE %s WITH (partitioned_by = ARRAY['linenumber']) AS " +
+                    "SELECT orderkey, linenumber FROM lineitem", table1));
+            queryRunner.execute(format("CREATE TABLE %s WITH (partitioned_by = ARRAY['orderstatus']) AS " +
+                    "SELECT orderkey, custkey, orderstatus FROM orders", table2));
+            queryRunner.execute(format("CREATE TABLE %s WITH (partitioned_by = ARRAY['nationkey']) AS " +
+                    "SELECT custkey, nationkey FROM customer", table3));
+
+            assertUpdate(createMV);
+            Assert.assertEquals(computeActual(joinSelect),
```

Comment:
We should do the plan validation as well to make sure it is indeed reading something from MV. Without it the query might as well read everything from base table and we would not know.